### PR TITLE
Do not verify block and use the RRS of the block before VRF shares meet the threshold

### PR DESCRIPTION
**Pull request by peterlimg**
Fixes: 
- #601 

Changes:
- Do not verify block if the round does not have VRF shares that could meet consensus threshold.
- Add VRF shares to the `verify block` message so that receiver could use them to verify and merge the VRF shares into local round, and then start to verify the block. Return error if the VRF shares are invalid.
- Do not send verify block message when received VRFShare.Round is < current round, send notarized block should be fine.

### CODE REVIEW

**Comment by dabasov** [comment]
I don't think we should resend VRFShare, synchronous protocol guarantees that we receive shares before the verify block, the only way we could miss it due to node offline, so it is not a problem for delayed replica to synchronize on notarization for this round.
The cons of this solution I'm worried about is network traffic, Imagine we have k generators and n miners with vrf threshold like 2/3. We will increase the size of traffic for verify round by 2/3*k*n*size(share) (we assume honest generators to have all threshold of shares)
Since we store full node structure in the vrf share, wonder why we can't store only miner key, traffic can grow tremendously.

I suppose at least to calculate message size of vrf share and figure out how much traffic we should to trade of

**Review by dabasov — CHANGES_REQUESTED** [review]
I don't think we should resend VRFShare, synchronous protocol guarantees that we receive shares before the verify block, the only way we could miss it due to node offline, so it is not a problem for delayed replica to synchronize on notarization for this round.
The cons of this solution I'm worried about is network traffic, Imagine we have k generators and n miners with vrf threshold like 2/3. We will increase the size of traffic for verify round by 2/3kn*size(share) (we assume honest generators to have all threshold of shares)
Since we store full node structure in the vrf share, wonder why we can't store only miner key, traffic can grow tremendously.

I suppose at least to calculate message size of vrf share and figure out how much traffic we should to trade of

**Comment by peterlimg** [comment]
@dabasov you're right, the network traffic would be a problem when nodes number increases. I've figured out why the chain is slow, there are mainly two places that can lead to slow collecting of VRF shares:
1. VRF shares could be dropped before they could be processed if the N2N sender request signature verification are failed due to CPU busy caused timeout. 
2. VRF shares will be ignored if it's previous round has no seed. 

So I did some tests to fix the above issues by
- Do not start the sender request signature verification immediately when the message is received, instead we can read out the data and do some basic check, and if it can not pass the check, we can ignore the message without doing the signature verification. In this way, we can save lot of CPU resources, and reduce the possibility that VRF shares are dropped due to the signature verification timeout error.
- Cache the received VRF shares for a round if the previous round has no seed yet, verify and collect those VRF shares once the previous round is ready. In this way, we don't need to wait for the re-sending of VRF shares.

So far, the fix looks good, check it [here](http://216.218.228.197:7087/_diagnostics).

File: code/go/0chain.net/miner/m_handler.go [hunk 1]
```diff
@@ -191,11 +191,12 @@ func VRFShareHandler(ctx context.Context, entity datastore.Entity) (
...
-		go func() {
-			mb.Miners.SendTo(ctx, VerifyBlockSender(hnb), found.ID)
-			mb.Miners.SendTo(ctx, MinerNotarizedBlockSender(hnb), found.ID)
-		}()
+		if err := node.ValidateSenderSignature(ctx); err != nil {
+			return nil, err
+		}
+
+		// send notarized block
+		go mb.Miners.SendTo(ctx, MinerNotarizedBlockSender(hnb), found.ID)
```
Comment (dabasov) [inline code/go/0chain.net/miner/m_handler.go:199]: I think it is not necessary to resend notarized block to delayed miner

**Review by dabasov — COMMENTED** [review]
LGTM

File: code/go/0chain.net/miner/m_handler.go (hunk 1 above)
Comment (dabasov) [reply code/go/0chain.net/miner/m_handler.go:199]: We can create an issue or discuss it in view-change

**Review by dabasov — APPROVED** [review]

**Pull request merged**